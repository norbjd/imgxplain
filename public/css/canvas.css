#canvases {
  position: relative;
}

#canvases canvas {
  position: absolute;
  top: 0;
  border: solid black;
}

#draw {
  z-index: 1;
}

#preview {
  z-index: 2;
}
