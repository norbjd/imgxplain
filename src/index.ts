import { App } from "./app";

window.addEventListener("DOMContentLoaded", () => {
  if (!window.AudioContext) {
    const errorMessage = "Web Audio not supported, change your browser";
    alert(errorMessage);
    throw new Error(errorMessage);
  }

  new App().run();
});
