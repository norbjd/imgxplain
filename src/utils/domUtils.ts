class DOMUtils {
  static getDocumentWidth(): number {
    return Math.max(
      document.body.scrollWidth,
      document.documentElement.scrollWidth,
      document.body.offsetWidth,
      document.documentElement.offsetWidth,
      document.documentElement.clientWidth
    );
  }

  static getDocumentHeight(): number {
    return Math.max(
      document.body.scrollHeight,
      document.documentElement.scrollHeight,
      document.body.offsetHeight,
      document.documentElement.offsetHeight,
      document.documentElement.clientHeight
    );
  }

  static getTranscriptDiv(): HTMLDivElement {
    return document.querySelector("#transcript") as HTMLDivElement;
  }

  static getMainCanvas(): HTMLCanvasElement {
    return document.querySelector("#draw") as HTMLCanvasElement;
  }

  static getPreviewCanvas(): HTMLCanvasElement {
    return document.querySelector("#preview") as HTMLCanvasElement;
  }

  static getMainDiv(): HTMLDivElement {
    return document.querySelector("#main") as HTMLDivElement;
  }

  static getWaveformDiv(): HTMLDivElement {
    return document.querySelector("#waveform");
  }

  static getEditorDiv(): HTMLDivElement {
    return document.querySelector("#editor");
  }

  static getEditForm(): HTMLFormElement {
    return document.forms.namedItem("edit");
  }

  static getLoadForm(): HTMLFormElement {
    return document.forms.namedItem("load");
  }

  static getAvailableMaxWidthForMainDiv(): number {
    return (
      (this.getDocumentWidth() -
        this.getMainDivMarginLeft() -
        this.getEditorDivWidthInPx()) *
      0.95
    );
  }

  static getAvailableMaxHeightForMainDiv(): number {
    return this.getDocumentHeight() - 60;
  }

  static getEditorDivWidthInPx(): number {
    const editorDivWidth = this.getEditorDiv()
      .style.width.replace("%", "")
      .replace("px", "");
    if (editorDivWidth == "0" || editorDivWidth == "") {
      return 0;
    }
    return (this.getDocumentWidth() * parseInt(editorDivWidth)) / 100 + 20;
  }

  static getMainDivMarginLeft(): number {
    return Number(this.getMainDiv().style.marginLeft.replace("px", ""));
  }

  static setHTMLInputElementValue(
    elementName: string,
    value: string | number
  ): void {
    (document.getElementById(elementName) as HTMLInputElement).value =
      "" + value;
  }

  static getPlayPauseButton(): HTMLButtonElement {
    return document.querySelector("[data-action='play']") as HTMLButtonElement;
  }

  static getTranscriptButton(): HTMLInputElement {
    return document.querySelector("#transcript_button");
  }

  static getCCButton(): HTMLInputElement {
    return document.querySelector("#cc");
  }

  static getWaveformZoomSlider(): HTMLInputElement {
    return document.querySelector("[data-action='zoom']");
  }

  static getExportActionsToJsonButton(): HTMLButtonElement {
    return document.querySelector(
      "[data-action='export']"
    ) as HTMLButtonElement;
  }

  static getExportToVideoButton(): HTMLButtonElement {
    return document.querySelector(
      "[data-action='export-video']"
    ) as HTMLButtonElement;
  }

  static getToggleEditorButton(): HTMLButtonElement {
    return document.getElementById("toggle_editor_button") as HTMLButtonElement;
  }

  static openEditor(): void {
    this.getToggleEditorButton().textContent = "ðŸ–‰ Close Editor";
    DOMUtils.getEditorDiv().style.width = "40%";
    DOMUtils.getEditorDiv().style.borderLeft = "1px solid black";
    DOMUtils.getEditorDiv().style.paddingLeft = "10px";
    DOMUtils.getEditorDiv().style.paddingRight = "10px";
  }

  static closeEditor(): void {
    this.getToggleEditorButton().textContent = "ðŸ–‰ Open Editor";
    DOMUtils.getEditorDiv().style.width = "0";
    DOMUtils.getEditorDiv().style.borderLeft = "";
    DOMUtils.getEditorDiv().style.paddingLeft = "0";
    DOMUtils.getEditorDiv().style.paddingRight = "0";
  }

  static removeLoadModalDialog(): void {
    document.getElementById("load_modal_dialog").remove();
  }
}

export { DOMUtils };
