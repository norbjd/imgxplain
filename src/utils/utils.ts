class Utils {
  static isEquivalent(a: unknown, b: unknown): boolean {
    if (typeof a != typeof b) {
      return false;
    }

    const aProps = Object.getOwnPropertyNames(a);
    const bProps = Object.getOwnPropertyNames(b);

    if (aProps.length != bProps.length) {
      return false;
    }

    for (let i = 0; i < aProps.length; i++) {
      const propName = aProps[i];
      if (a[propName] !== b[propName]) {
        return false;
      }
    }

    return true;
  }
}

export { Utils };
