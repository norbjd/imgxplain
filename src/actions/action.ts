import { Canvas } from "../canvas";

class Action<T extends ActionOptions> {
  canvas: Canvas;
  type: string;
  options: T;

  constructor(canvas: Canvas, type: string, options: T) {
    this.canvas = canvas;
    this.type = type;
    this.options = options;
  }

  do(context: CanvasRenderingContext2D): void {
    throw new Error("Not implemented");
  }

  execute(): void { this.do(this.canvas.mainContext); }
  preview(): void { this.do(this.canvas.previewContext); }
}

class ActionOptions {}

export {
  Action,
  ActionOptions,
};
