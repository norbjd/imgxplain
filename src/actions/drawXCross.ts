import { Action, ActionOptions } from "./action";
import { Canvas } from "../canvas";

class DrawXCrossOptions extends ActionOptions {
  centerX: number;
  centerY: number;
  radius: number;
  lineWidth: number;
  color: string;

  constructor(
    centerX: number,
    centerY: number,
    radius: number,
    lineWidth: number,
    color: string
  ) {
    super();
    this.centerX = centerX;
    this.centerY = centerY;
    this.radius = radius;
    this.lineWidth = lineWidth;
    this.color = color;
  }
}

class DrawXCross extends Action<DrawXCrossOptions> {
  constructor(canvas: Canvas, options: DrawXCrossOptions) {
    super(canvas, "draw_xcross", options);
  }

  do(context: CanvasRenderingContext2D): void {
    const xCrossCanvasCenter = this.canvas.getPointProjectionOnCanvas(
      this.options.centerX,
      this.options.centerY
    );
    const xCrossCanvasCenterX = xCrossCanvasCenter.x;
    const xCrossCanvasCenterY = xCrossCanvasCenter.y;

    const ratio = this.canvas.getImageRatio();
    const xCrossImgRadiusX = this.options.radius / ratio.ratioWidth;
    const xCrossImgRadiusY = this.options.radius / ratio.ratioHeight;

    context.save();

    context.translate(xCrossCanvasCenterX, xCrossCanvasCenterY);
    context.rotate(Math.PI / 4);
    context.translate(-xCrossCanvasCenterX, -xCrossCanvasCenterY);

    context.lineWidth = this.options.lineWidth;
    context.strokeStyle = this.options.color;

    context.beginPath();
    context.moveTo(xCrossCanvasCenterX - xCrossImgRadiusX, xCrossCanvasCenterY);
    context.lineTo(xCrossCanvasCenterX + xCrossImgRadiusX, xCrossCanvasCenterY);
    context.stroke();
    context.moveTo(xCrossCanvasCenterX, xCrossCanvasCenterY - xCrossImgRadiusY);
    context.lineTo(xCrossCanvasCenterX, xCrossCanvasCenterY + xCrossImgRadiusY);
    context.stroke();

    context.restore();
  }
}

export { DrawXCross, DrawXCrossOptions };
