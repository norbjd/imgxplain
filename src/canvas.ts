import { Action, ActionOptions } from "./actions/action";
import { Zoom, ZoomOptions } from "./actions/zoom";
import { DOMUtils } from "./utils/domUtils";

class Canvas {
  img: HTMLImageElement;
  imgTopLeftX: number;
  imgTopLeftY: number;
  imgBottomRightX: number;
  imgBottomRightY: number;

  mainCanvas: HTMLCanvasElement;
  mainContext: CanvasRenderingContext2D;

  previewCanvas: HTMLCanvasElement;
  previewContext: CanvasRenderingContext2D;

  clickPosition: { x: number; y: number };

  ccEnabled: boolean;

  constructor(imgBlob: string) {
    this.img = new Image();
    this.img.src = imgBlob;
    this.img.addEventListener("load", () => {
      this.renderImage();
    });

    this.imgTopLeftX = 0;
    this.imgTopLeftY = 0;
    this.imgBottomRightX = this.img.naturalWidth;
    this.imgBottomRightY = this.img.naturalHeight;

    this.mainCanvas = DOMUtils.getMainCanvas();
    this.mainContext = this.mainCanvas.getContext("2d");

    this.previewCanvas = DOMUtils.getPreviewCanvas();
    this.previewContext = this.previewCanvas.getContext("2d");

    this.clickPosition = { x: 0, y: 0 };

    this.ccEnabled = false;
  }

  render(
    topLeftX: number,
    topLeftY: number,
    bottomRightX: number,
    bottomRightY: number
  ): void {
    this.imgTopLeftX = topLeftX;
    this.imgTopLeftY = topLeftY;
    this.imgBottomRightX = bottomRightX;
    this.imgBottomRightY = bottomRightY;

    const imgWidth = this.imgBottomRightX - this.imgTopLeftX;
    const imgHeight = this.imgBottomRightY - this.imgTopLeftY;

    const imgRatio = imgWidth / imgHeight;

    const mainDivMaxWidth = DOMUtils.getAvailableMaxWidthForMainDiv();
    const mainDivMaxHeight = DOMUtils.getAvailableMaxHeightForMainDiv();

    let canvasWidth = Math.min(mainDivMaxWidth, imgWidth);
    let canvasHeight = canvasWidth / imgRatio;

    if (canvasHeight >= mainDivMaxHeight) {
      canvasHeight = mainDivMaxHeight;
      canvasWidth = canvasHeight * imgRatio;
    }

    this.mainCanvas.width = canvasWidth;
    this.mainCanvas.height = canvasHeight;
    this.mainContext.drawImage(
      this.img,
      this.imgTopLeftX,
      this.imgTopLeftY,
      imgWidth,
      imgHeight,
      0,
      0,
      canvasWidth,
      canvasHeight
    );

    this.previewCanvas.width = canvasWidth;
    this.previewCanvas.height = canvasHeight;
    this.previewCanvas.style.display = "none";

    // TODO : move somewhere else
    // -20 : padding
    DOMUtils.getTranscriptDiv().style.width = "" + (canvasWidth + 5 - 20);
  }

  renderFullImage(): void {
    this.render(0, 0, this.img.naturalWidth, this.img.naturalHeight);
  }

  renderImage(): void {
    this.render(
      this.imgTopLeftX,
      this.imgTopLeftY,
      this.imgBottomRightX,
      this.imgBottomRightY
    );
  }

  getWidthAndHeight(): { width: number; height: number } {
    return {
      width: this.mainCanvas.width,
      height: this.mainCanvas.height,
    };
  }

  getImageRatio(): { ratioWidth: number; ratioHeight: number } {
    const imgWidth = this.imgBottomRightX - this.imgTopLeftX;
    const imgHeight = this.imgBottomRightY - this.imgTopLeftY;

    const canvasWidthAndHeight = this.getWidthAndHeight();

    return {
      ratioWidth: imgWidth / canvasWidthAndHeight.width,
      ratioHeight: imgHeight / canvasWidthAndHeight.height,
    };
  }

  getPointProjectionOnCanvas(
    imgPointX: number,
    imgPointY: number
  ): { x: number; y: number } {
    const ratio = this.getImageRatio();

    const canvasPointX = (imgPointX - this.imgTopLeftX) / ratio.ratioWidth;
    const canvasPointY = (imgPointY - this.imgTopLeftY) / ratio.ratioHeight;

    return {
      x: canvasPointX,
      y: canvasPointY,
    };
  }

  getCursorPositionOnCanvas(event: MouseEvent): { x: number; y: number } {
    const rect = this.mainCanvas.getBoundingClientRect();
    const x = event.clientX - rect.left;
    const y = event.clientY - rect.top;

    const ratio = this.getImageRatio();

    return {
      x: Math.floor(x * ratio.ratioWidth) + this.imgTopLeftX,
      y: Math.floor(y * ratio.ratioHeight) + this.imgTopLeftY,
    };
  }

  addClickEventListener(listener: (event: MouseEvent) => void): void {
    this.mainCanvas.addEventListener("click", listener);
  }

  executeAction(action: Action<ActionOptions>): void {
    action.execute();
  }

  executeActions(actions: Action<ActionOptions>[]): void {
    if (actions.find((action) => action instanceof Zoom) == undefined) {
      this.renderFullImage();
    }

    this.renderImage();
    actions.forEach((action) => {
      this.executeAction(action);
    });
  }

  previewAction(action: Action<ActionOptions>, timeInMs: number): void {
    this.previewCanvas.style.display = "block";
    action.preview();
    setTimeout(() => {
      this.previewContext.clearRect(
        0,
        0,
        this.previewCanvas.width,
        this.previewCanvas.height
      );
      this.previewCanvas.style.display = "none";
      if (action instanceof Zoom) {
        const unzoom = new Zoom(
          this,
          new ZoomOptions(0, 0, this.img.naturalWidth, this.img.naturalHeight)
        );
        this.executeAction(unzoom);
      }
    }, timeInMs);
  }
}

export { Canvas };
