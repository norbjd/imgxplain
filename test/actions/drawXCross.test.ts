import { expect, it, jest, beforeEach } from "@jest/globals";
import { DrawXCross, DrawXCrossOptions } from "../../src/actions/drawXCross";
import { Canvas } from "../../src/canvas";
import { DOMUtils } from "../../src/utils/domUtils";

import { readFileSync } from "fs";
import { resolve } from "path";
import { initCanvas } from "./utils";

const html = readFileSync(
  resolve(__dirname, "../../public/index.html"),
  "utf8"
);
jest.dontMock("fs");
document.documentElement.innerHTML = html.toString();

let canvas: Canvas;
beforeEach(() => {
  canvas = initCanvas();
});

it("should draw a red x cross in the middle", () => {
  new DrawXCross(
    canvas,
    new DrawXCrossOptions(150, 75, 50, 5, "#ff0000")
  ).execute();
  const imgBase64 = DOMUtils.getMainCanvas().toDataURL("image/png");
  expect(imgBase64).toMatchSnapshot();
});

it("should draw a blue x cross", () => {
  new DrawXCross(
    canvas,
    new DrawXCrossOptions(50, 70, 30, 3, "#0000ff")
  ).execute();
  const imgBase64 = DOMUtils.getMainCanvas().toDataURL("image/png");
  expect(imgBase64).toMatchSnapshot();
});

it("should preview a red x cross in the middle", () => {
  new DrawXCross(
    canvas,
    new DrawXCrossOptions(150, 75, 50, 5, "#ff0000")
  ).preview();
  const imgBase64 = DOMUtils.getPreviewCanvas().toDataURL("image/png");
  expect(imgBase64).toMatchSnapshot();
});

it("should preview a blue x cross", () => {
  new DrawXCross(
    canvas,
    new DrawXCrossOptions(50, 70, 30, 3, "#0000ff")
  ).preview();
  const imgBase64 = DOMUtils.getPreviewCanvas().toDataURL("image/png");
  expect(imgBase64).toMatchSnapshot();
});
