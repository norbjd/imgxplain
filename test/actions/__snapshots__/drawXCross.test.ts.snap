// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should draw a blue x cross 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAEcUlEQVR4nO3dvWpUWxjH4dfRXvEOgiQYP8BGULAQvJIEbfSGBEWvRUhjqRIRPxAVLGxSC3HmFMOQxIRzosnstf77PE+3d4pZzftjhZm99pnZbDYrgACT1gsAOC7BAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwgRpNgff5ctbvb4pOPZzqt2tlpvQrgd4MH6+3bqvX1qitX+ozWbFb16FHV7dtV37+3Xg2w35nZbDYb6sM+faq6erXq58/59dpa1fZ21blzQ63g302nVQ8eVD17Nr9eX6/a2qq6eLHtuoC5QXdYKytVly7tXb9/389Oa7GzWsSqqurWraoLF9qtCTho0GBNJlVv3swjtfDhQ/tozWZVDx9WPX68d29jo+rp0/magT4MPo69RUusIEeTkewlWmIFWZqNZetoiRXkaTqaraIlVpCp+XgOHS2xglxdjOhQ0RIryNbNmC47WmIF+boa1WVFS6xgHLob19OOlljBeHQ5sqcVLbGCcel2bE8aLbGC8el6dP82WmIF49T9+P5ptMQKxitihI8bLbGCcRv0AL+Tmk6rrl+fn1q6sLo6vz57Vqxg7KKCVVX161fVtWtV797t3Vtdrbpzp+r58717m5tVT56IFYxJXLCqjt5p7WdnBeMUOdKTSdXr11WXLx/+m1jBeMWO9WRSdffu4ftbW/MdGDA+kcE66tvAhR7OiAeWIy5Y02nV/fsHY7W5efDfQ9GCcYoK1lGv4trYmH8buL3d/ox4YLligvVfPwptfUY8sHwRwTruL9hFC8at+2D96eM2ogXj1XWw/vbZQNGCceo2WCd9kFm0YHy6DNZpnbogWjAu3QXrtI+IES0Yj66CtazzrEQLxqGbYC378D3RgnxdBGuok0JFC7I1D9bQxxqLFuRqGqxWZ7CLFmRqFqzWL4wQLcjTJFitY7UgWpBl8GD1EqsF0YIcgyait1gtiBZkGDQTOztVL17sXW9uto/VwmRS9epV1dra3r2vX6u+fGm3JuCgwV/z9eNH1b17VTdv9hOr/RavEPv4serly6obN1qvCFho8l7CnZ2q8+f7i9XC7m7Vt29VKyutVwLsF/kiVeD/qdM9DsBhggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhDjHwYaEyMv61sqAAAAAElFTkSuQmCC"`;

exports[`should draw a red x cross in the middle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAFbElEQVR4nO3dvWoUbRTA8ZMvCCJWQgJpLIxgrsFGIWAI6pW8lRdhY5vWezAiCoGk8B4sTG0hdhYSMHHeYh2M+dhsdic7e57n9yszzWn2z5nJfMw1TdMEQALzfQ8AMCrBAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hCs0hwd9T3B7Dg+7nsCOiZYJdnZibhzJ+L9+74n6d/hYcTGRsT+ft+T0KG5pmmavoegAzs7Ef/9F9E0EfPzEbu7EdvbfU/Vj8PDiMePI75+jbh1K+Ldu4gnT/qeig7YsEpwOlYREb9/Rzx/XuemdTpWERE/f0Y8e2bTKoQNK7ujo8Fp4K9f548tLER8+BCxuTn9ufpwNlanra9HfP4csbg4/bnojA0ru+XlwSa1sHD+2MlJxNOndWxaw2K1uhrx9q1YFUCwSrC5OdikLopWDaeHV8Vqfz/i4cPpz0XnBKsUtUZLrKoiWCWpLVpiVR3BKk0t0RKrKglWiUqPllhVS7BKVWq0xKpqglWy0qIlVtUTrNKVEi2xIgSrDtmjJVb8IVi1yBotseIUwapJtmiJFWcIVm2yREusuIBg1WjWoyVWXEKwajWr0RIrhhCsms1atMSKKwhW7WYlWmLFCASL/qMlVoxIsBjoK1pixTUIFn9NO1pixTUJFv+aVrTEijEIFufddLTEijEJFhe7qWiJFRMQLC7XdbTEigkJFsN1FS2xogOCxdUmjZZY0RHBYjTjRkus6JBgMbrrRkus6Nhc0zRN30OQzN5exNZWxMnJ+WPz8xG7uxEPHogVnRMsxjMsWgsLEXfvRnz7dv7Y2lrEwUHE+vrNz0hxBIvxDYvWRWxWTMg1LMY37JrWWWJFB2xYTO6qTWtlJeLTJ6eBTMyGxeTu3Rtcs7rM9+8RX75Mbx6KJVhMpr114aIL7K1Z+LAFRRAsxjfsPquzRIsOCBbjGRarlZXLby598SLi48ebn48iCRbXd9Ud7AcHl//38OQkYnvbpsVYBIvrGfVxm74/bEGRBIvRXffZQNGiY4LFaMZ9kFm06JBgcbVJ37ogWnREsBiuq1fEiBYdECwu1/X7rESLCQkWF7upl++JFhMQLM676TeFihZjEiz+Na3XGosWYxAs/pr2O9hFi2sSLAb6+mCEaHENgkX/X7cRLUYkWLXrO1Yt0WIEglWzWYlVS7S4gmDVatZi1RIthhCsGs1qrFqixSUEqzazHquWaHEBwapJlli1RIszBKsW2WLVEi1OEawaZI1VS7T4Q7BKlz1WLdEiBKtspcSqJVrVE6xSlRarlmhVTbBKVGqsWqJVLcEqTemxaolWlQSrJLXEqiVa1RGsUtQWq5ZoVUWwSlBrrFqiVY25pmmavodgAsfHERsbg2idtbYWcXAQsb4+/bn6sLcXsbUVcXJy/tjSUsSPHxHLy9Ofi87YsLJbXIx48ybi9u1//766OvgB1xKriOGb1qtXYlUAwSrBo0eDH2obrRpOAy9zUbRev454+bK/meiMYJWijdb9+/XGqtVGa2lJrArjGlZpjo8Hp4lEHB05DSyMYAFpOCUE0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgjf8BWTdqFvla8cMAAAAASUVORK5CYII="`;

exports[`should preview a blue x cross 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAEcUlEQVR4nO3dvWpUWxjH4dfRXvEOgiQYP8BGULAQvJIEbfSGBEWvRUhjqRIRPxAVLGxSC3HmFMOQxIRzosnstf77PE+3d4pZzftjhZm99pnZbDYrgACT1gsAOC7BAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwgRpNgff5ctbvb4pOPZzqt2tlpvQrgd4MH6+3bqvX1qitX+ozWbFb16FHV7dtV37+3Xg2w35nZbDYb6sM+faq6erXq58/59dpa1fZ21blzQ63g302nVQ8eVD17Nr9eX6/a2qq6eLHtuoC5QXdYKytVly7tXb9/389Oa7GzWsSqqurWraoLF9qtCTho0GBNJlVv3swjtfDhQ/tozWZVDx9WPX68d29jo+rp0/magT4MPo69RUusIEeTkewlWmIFWZqNZetoiRXkaTqaraIlVpCp+XgOHS2xglxdjOhQ0RIryNbNmC47WmIF+boa1WVFS6xgHLob19OOlljBeHQ5sqcVLbGCcel2bE8aLbGC8el6dP82WmIF49T9+P5ptMQKxitihI8bLbGCcRv0AL+Tmk6rrl+fn1q6sLo6vz57Vqxg7KKCVVX161fVtWtV797t3Vtdrbpzp+r58717m5tVT56IFYxJXLCqjt5p7WdnBeMUOdKTSdXr11WXLx/+m1jBeMWO9WRSdffu4ftbW/MdGDA+kcE66tvAhR7OiAeWIy5Y02nV/fsHY7W5efDfQ9GCcYoK1lGv4trYmH8buL3d/ox4YLligvVfPwptfUY8sHwRwTruL9hFC8at+2D96eM2ogXj1XWw/vbZQNGCceo2WCd9kFm0YHy6DNZpnbogWjAu3QXrtI+IES0Yj66CtazzrEQLxqGbYC378D3RgnxdBGuok0JFC7I1D9bQxxqLFuRqGqxWZ7CLFmRqFqzWL4wQLcjTJFitY7UgWpBl8GD1EqsF0YIcgyait1gtiBZkGDQTOztVL17sXW9uto/VwmRS9epV1dra3r2vX6u+fGm3JuCgwV/z9eNH1b17VTdv9hOr/RavEPv4serly6obN1qvCFho8l7CnZ2q8+f7i9XC7m7Vt29VKyutVwLsF/kiVeD/qdM9DsBhggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhBDsIAYggXEECwghmABMQQLiCFYQAzBAmIIFhDjHwYaEyMv61sqAAAAAElFTkSuQmCC"`;

exports[`should preview a red x cross in the middle 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAFbElEQVR4nO3dvWoUbRTA8ZMvCCJWQgJpLIxgrsFGIWAI6pW8lRdhY5vWezAiCoGk8B4sTG0hdhYSMHHeYh2M+dhsdic7e57n9yszzWn2z5nJfMw1TdMEQALzfQ8AMCrBAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hCs0hwd9T3B7Dg+7nsCOiZYJdnZibhzJ+L9+74n6d/hYcTGRsT+ft+T0KG5pmmavoegAzs7Ef/9F9E0EfPzEbu7EdvbfU/Vj8PDiMePI75+jbh1K+Ldu4gnT/qeig7YsEpwOlYREb9/Rzx/XuemdTpWERE/f0Y8e2bTKoQNK7ujo8Fp4K9f548tLER8+BCxuTn9ufpwNlanra9HfP4csbg4/bnojA0ru+XlwSa1sHD+2MlJxNOndWxaw2K1uhrx9q1YFUCwSrC5OdikLopWDaeHV8Vqfz/i4cPpz0XnBKsUtUZLrKoiWCWpLVpiVR3BKk0t0RKrKglWiUqPllhVS7BKVWq0xKpqglWy0qIlVtUTrNKVEi2xIgSrDtmjJVb8IVi1yBotseIUwapJtmiJFWcIVm2yREusuIBg1WjWoyVWXEKwajWr0RIrhhCsms1atMSKKwhW7WYlWmLFCASL/qMlVoxIsBjoK1pixTUIFn9NO1pixTUJFv+aVrTEijEIFufddLTEijEJFhe7qWiJFRMQLC7XdbTEigkJFsN1FS2xogOCxdUmjZZY0RHBYjTjRkus6JBgMbrrRkus6Nhc0zRN30OQzN5exNZWxMnJ+WPz8xG7uxEPHogVnRMsxjMsWgsLEXfvRnz7dv7Y2lrEwUHE+vrNz0hxBIvxDYvWRWxWTMg1LMY37JrWWWJFB2xYTO6qTWtlJeLTJ6eBTMyGxeTu3Rtcs7rM9+8RX75Mbx6KJVhMpr114aIL7K1Z+LAFRRAsxjfsPquzRIsOCBbjGRarlZXLby598SLi48ebn48iCRbXd9Ud7AcHl//38OQkYnvbpsVYBIvrGfVxm74/bEGRBIvRXffZQNGiY4LFaMZ9kFm06JBgcbVJ37ogWnREsBiuq1fEiBYdECwu1/X7rESLCQkWF7upl++JFhMQLM676TeFihZjEiz+Na3XGosWYxAs/pr2O9hFi2sSLAb6+mCEaHENgkX/X7cRLUYkWLXrO1Yt0WIEglWzWYlVS7S4gmDVatZi1RIthhCsGs1qrFqixSUEqzazHquWaHEBwapJlli1RIszBKsW2WLVEi1OEawaZI1VS7T4Q7BKlz1WLdEiBKtspcSqJVrVE6xSlRarlmhVTbBKVGqsWqJVLcEqTemxaolWlQSrJLXEqiVa1RGsUtQWq5ZoVUWwSlBrrFqiVY25pmmavodgAsfHERsbg2idtbYWcXAQsb4+/bn6sLcXsbUVcXJy/tjSUsSPHxHLy9Ofi87YsLJbXIx48ybi9u1//766OvgB1xKriOGb1qtXYlUAwSrBo0eDH2obrRpOAy9zUbRev454+bK/meiMYJWijdb9+/XGqtVGa2lJrArjGlZpjo8Hp4lEHB05DSyMYAFpOCUE0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgDcEC0hAsIA3BAtIQLCANwQLSECwgjf8BWTdqFvla8cMAAAAASUVORK5CYII="`;
